src/t_set.c

```diff
@@ -963,18 +1055,20 @@ void sinterGenericCommand(client *c, robj **setkeys,
                 if (limit && cardinality >= limit)
                     break;
             } else if (!dstkey) {
-                if (encoding == OBJ_ENCODING_HT)
-                    addReplyBulkCBuffer(c,elesds,sdslen(elesds));
+                if (str != NULL)
+                    addReplyBulkCBuffer(c, str, len);
                 else
                     addReplyBulkLongLong(c,intobj);
                 cardinality++;
             } else {
-                if (encoding == OBJ_ENCODING_INTSET) {
-                    elesds = sdsfromlonglong(intobj);
+                if (encoding == OBJ_ENCODING_HT) {
+                    sds elesds = (sds)str;
                     setTypeAdd(dstset,elesds);
-                    sdsfree(elesds);
                 } else {
-                    setTypeAdd(dstset,elesds);
+                    sds elesds = (str != NULL) ?
+                        sdsnewlen(str, len) : sdsfromlonglong(intobj);
+                    setTypeAdd(dstset, elesds);
+                    sdsfree(elesds);
```

Comment:
maybe we can avoid the extra copy and free if we add a setTypeAdd variant that takes char* and size_t.
i.e. specifically for listpack insertion it's a waste.

Reply:
Done, and a variant for adding ints.

Variants of `setTypeIsMember()` and `setTypeRemove()` would also allow skipping some temporary sds values (but I'll handle the other CR comments first).